# Complete XGBoost Track

## Stage 4I — Initial XGBoost Model and Importance Analysis

### 0. Stage 4I Objective

Why: Set the scope before any result is shown.

The code below reads compact saved evidence.

Interpretation: The track starts with non-sensitive Discovery rows and keeps Test locked.

Conclusion: Stage 4I can begin without using Test evidence.

In [1]:
print('Stage 4I uses only saved non-sensitive Discovery rows. The locked Test Set stays unused.')

Stage 4I uses only saved non-sensitive Discovery rows. The locked Test Set stays unused.


### 1. Imports and Configuration

Why: Load only configuration and saved evidence.

The code below reads compact saved evidence.

Interpretation: XGBoost 3.3.0 and SHAP 0.52.0 are available; cache-only state is explicit.

Conclusion: The saved environment can read every XGBoost artifact.

In [2]:
from pathlib import Path
import json, os, pandas as pd
from IPython.display import display
ROOT=Path.cwd()
CACHE_ONLY=os.getenv('STAGE4_XGBOOST_CACHE_ONLY','0')=='1'
def show_json(rel):
    value=json.loads((ROOT/rel).read_text(encoding='utf-8')); display(value); return value
def show_csv(rel,n=10):
    value=pd.read_csv(ROOT/rel); display(value.head(n)); return value
preflight=show_json('artifacts/reports/stage4i_preflight.json')

{'stage': 'stage4i',
 'created_at_utc': '2026-07-14T19:04:06.085237+00:00',
 'prior_stages': {'stage1': {'path': 'artifacts/reports/prompt1_verification.json',
   'status': 'PASS',
   'sha256': '50172f0f3767f2326f3ab626f25bd483f69c7f625ee847d82f4e581495aeaafb'},
  'stage2': {'path': 'artifacts/reports/prompt2_verification.json',
   'status': 'PASS',
   'sha256': '38be7bec7cb1b44227b0a5ebfcd7b099b1905d4bd92527eb9e237a78c8c530eb'},
  'stage3': {'path': 'artifacts/reports/stage3_verification.json',
   'status': 'PASS',
   'sha256': '119b86b0d6119c9e4359c1acf4dfac69d6882e5df9ee744b928db43fe4a72e20'},
  'stage4a': {'path': 'artifacts/reports/stage4a_verification.json',
   'status': 'PASS',
   'sha256': '60c2c8d4b5dfc5304109d04b72b4a5255f0e2cafa6a75cacf7e8ce5e61038edd'},
  'stage4b': {'path': 'artifacts/reports/stage4b_verification.json',
   'status': 'PASS',
   'sha256': '55daf623c5ff33b60711dca8f7b0a78d549551d1ea41f370be3f76b65802bd8e'},
  'stage4c': {'path': 'artifacts/reports/stage4c_ver

### 2. Previous-Stage Validation

Why: Confirm that every required earlier Stage passed.

The code below reads compact saved evidence.

Interpretation: Stages 1 through 4H all report PASS.

Conclusion: The XGBoost track has a valid prior-stage handoff.

In [3]:
display(pd.DataFrame(preflight['prior_stages']).T[['path','status']])

,path,status
stage1,artifacts/reports/prompt1_verification.json,PASS
stage2,artifacts/reports/prompt2_verification.json,PASS
stage3,artifacts/reports/stage3_verification.json,PASS
stage4a,artifacts/reports/stage4a_verification.json,PASS
stage4b,artifacts/reports/stage4b_verification.json,PASS
stage4c,artifacts/reports/stage4c_verification.json,PASS
stage4de,artifacts/reports/stage4de_verification.json,PASS
stage4f,artifacts/reports/stage4f_gate_verification.json,PASS
stage4g,artifacts/reports/stage4g_gate_verification.json,PASS
stage4h,artifacts/reports/stage4h_verification.json,PASS


### 3. Protected File Check

Why: Confirm immutable inputs did not change.

The code below reads compact saved evidence.

Interpretation: All 985 protected files match the baseline.

Conclusion: Source data, splits, prior notebooks, and prior artifacts stayed unchanged.

In [4]:
show_json('artifacts/manifests/stage4/xgboost/stage4k_protected_hashes_after.json')

{'file_count': 985, 'mismatches': {}, 'status': 'PASS'}

{'file_count': 985, 'mismatches': {}, 'status': 'PASS'}

### 4. Discovery Sample

Why: Show the fixed 50,000/15,000 Discovery roles.

The code below reads compact saved evidence.

Interpretation: Discovery has 50,000 Train and 15,000 Validation rows with zero Test overlap.

Conclusion: Stage 4I used the fixed sample exactly.

In [5]:
samples=show_json('artifacts/splits/stage4/stage4_sample_verification.json'); display(samples['samples']['discovery'])

{'stage': 'stage4a',
 'random_seed': 42,
 'sampling_method': 'one seeded per-bin shuffle with largest-remainder proportional allocation',
 'target_bin_source': 'saved split_config.json bin_edges',
 'source_train_rows': 399788,
 'locked_test_rows': 99948,
 'total_sample_rows': 290000,
 'all_sample_rows_unique': True,
 'all_rows_in_saved_train': True,
 'test_overlap_rows': 0,
 'samples': {'discovery': {'rows': 65000,
   'train_rows': 50000,
   'validation_rows': 15000,
   'expected_rows': 65000,
   'row_ids_unique': True,
   'maximum_target_bin_proportion_difference': 8.993074021537706e-06,
   'role_maximum_target_bin_proportion_difference': {'train': 1.2069997098462792e-05,
    'validation': 3.2547250042522236e-05},
   'sha256': '79cf8051552f1edc1e68b9284d0a3ccfb49f5ceb3afd4b767a0da5dee372320e',
   'valid': True},
  'feature_confirmation': {'rows': 100000,
   'train_rows': 80000,
   'validation_rows': 20000,
   'expected_rows': 100000,
   'row_ids_unique': True,
   'maximum_target_bin_p

{'rows': 65000,
 'train_rows': 50000,
 'validation_rows': 15000,
 'expected_rows': 65000,
 'row_ids_unique': True,
 'maximum_target_bin_proportion_difference': 8.993074021537706e-06,
 'role_maximum_target_bin_proportion_difference': {'train': 1.2069997098462792e-05,
  'validation': 3.2547250042522236e-05},
 'sha256': '79cf8051552f1edc1e68b9284d0a3ccfb49f5ceb3afd4b767a0da5dee372320e',
 'valid': True}

### 5. CPU and GPU Execution Mode

Why: Show why CPU with two threads was frozen.

The code below reads compact saved evidence.

Interpretation: CPU passed; requested CUDA fell back to CPU, so two-thread CPU mode was frozen.

Conclusion: Real fits did not use an unvalidated GPU path.

In [6]:
display({'selected':preflight['selected_execution_mode'],'threads':preflight['thread_count'],'smoke':preflight['smoke']})

{'selected': 'cpu',
 'threads': 2,
 'smoke': {'cpu': {'status': 'PASS',
   'device': 'cpu',
   'fit_seconds': 0.2649310999986483,
   'finite_predictions': True,
   'prediction_rows': 1000,
   'sparse_output': True,
   'transformed_shape': [100, 148],
   'transformed_bytes': 40820},
  'gpu': {'status': 'FAIL',
   'device': 'cuda',
   'fit_seconds': 0.17021750001003966,
   'finite_predictions': True,
   'prediction_rows': 1000,
   'sparse_output': True,
   'transformed_shape': [100, 148],
   'transformed_bytes': 40820,
   'observed_fallback_to_cpu': True,
   'error': 'XGBoost warned that no visible GPU was found and changed device from GPU to CPU.'}}}

### 6. XGBoost Feature Packs

Why: Show the two eligible Stage 4I packs.

The code below reads compact saved evidence.

Interpretation: The base pack has controlled one-hot Features; the sparse pack also adds fixed Features and training-fit frequencies.

Conclusion: Only the two eligible Stage 4I packs were compared.

In [7]:
packs=show_json('artifacts/features/stage4/boosting_feature_packs.json'); display({k:packs['packs'][k] for k in ['boosting_base_v1','xgboost_sparse_v1']})

{'version': 'stage4b_initial_boosting_packs_v1_20260714',
 'selection_basis': 'fixed design from non-sensitive training metadata and prior training-only evidence',
 'target_encoding_used': False,
 'sensitive_interactions_used': False,
 'packs': {'boosting_base_v1': {'numeric': ['applicant_income_000s',
    'population',
    'hud_median_family_income',
    'number_of_owner_occupied_units',
    'number_of_1_to_4_family_units',
    'applicant_income_to_area_income',
    'tract_income_ratio',
    'owner_occupied_unit_ratio',
    'family_units_per_1000_people',
    'owner_occupied_units_per_1000_people',
    'has_co_applicant'],
   'categorical': ['agency_name',
    'loan_type_name',
    'property_type_name',
    'loan_purpose_name',
    'owner_occupancy_name',
    'preapproval_name',
    'state_name',
    'lien_status_name',
    'loan_program_group',
    'applicant_income_area_group',
    'tract_income_level',
    'us_region'],
   'raw': ['applicant_income_000s',
    'population',
    'hud

{'boosting_base_v1': {'numeric': ['applicant_income_000s',
   'population',
   'hud_median_family_income',
   'number_of_owner_occupied_units',
   'number_of_1_to_4_family_units',
   'applicant_income_to_area_income',
   'tract_income_ratio',
   'owner_occupied_unit_ratio',
   'family_units_per_1000_people',
   'owner_occupied_units_per_1000_people',
   'has_co_applicant'],
  'categorical': ['agency_name',
   'loan_type_name',
   'property_type_name',
   'loan_purpose_name',
   'owner_occupancy_name',
   'preapproval_name',
   'state_name',
   'lien_status_name',
   'loan_program_group',
   'applicant_income_area_group',
   'tract_income_level',
   'us_region'],
  'raw': ['applicant_income_000s',
   'population',
   'hud_median_family_income',
   'number_of_owner_occupied_units',
   'number_of_1_to_4_family_units',
   'applicant_income_to_area_income',
   'tract_income_ratio',
   'owner_occupied_unit_ratio',
   'family_units_per_1000_people',
   'owner_occupied_units_per_1000_people',


### 7. Serializable XGBoost Bundle

Why: Confirm sparse serialization before real Candidates.

The code below reads compact saved evidence.

Interpretation: The 1,000/500 bundle smoke stayed sparse and survived joblib reload.

Conclusion: The custom eval-set bundle was safe before full Candidates.

In [8]:
show_json('artifacts/reports/stage4i_bundle_smoke.json')

{'stage': 'stage4i',
 'fit_rows': 1000,
 'validation_rows': 500,
 'feature_count': 138,
 'sparse_train': True,
 'best_iteration': 8,
 'mae': 120.93219631958007,
 'finite_predictions': True,
 'joblib_roundtrip': True,
 'status': 'PASS'}

{'stage': 'stage4i',
 'fit_rows': 1000,
 'validation_rows': 500,
 'feature_count': 138,
 'sparse_train': True,
 'best_iteration': 8,
 'mae': 120.93219631958007,
 'finite_predictions': True,
 'joblib_roundtrip': True,
 'status': 'PASS'}

### 8. Three Required Candidates

Why: Show the three required comparisons.

The code below reads compact saved evidence.

Interpretation: The required base/raw, sparse/raw, and sparse/log1p Candidates all completed.

Conclusion: Pack and target-mode questions were answered within budget.

In [9]:
candidates=show_csv('artifacts/results/stage4/xgboost/initial/stage4i_candidate_results.csv',4); display(candidates.iloc[:3])

,experiment_id,feature_pack,target_mode,max_depth,best_iteration,fit_seconds,mae,mse,rmse,mape_percent,...,negative_prediction_rate,mae_usd,rmse_usd,original_scale,top_decile_mae,top_five_percent_mae,underestimation_rate,overestimation_rate,training_mae,validation_train_mae_gap_percent
0,stage4i__candidate_03_sparse_log1p,xgboost_sparse_v1,log1p,6,720,18.142525,64.685719,19237.585130,138.699622,33.859754,...,0.000000,64685.719242,138699.621953,True,214.398433,306.327287,0.540400,0.459600,55.623919,16.291193
1,stage4i__candidate_04_safer_depth5_log1p,xgboost_sparse_v1,log1p,5,1132,26.380924,64.793724,19005.618345,137.860866,33.925681,...,0.000000,64793.723688,137860.865894,True,214.307146,305.693304,0.539200,0.460800,56.445130,14.790636
2,stage4i__candidate_02_sparse_raw,xgboost_sparse_v1,raw,6,688,17.137593,66.430050,18523.315288,136.100387,43.024905,...,0.007267,66430.050314,136100.386804,True,202.129636,287.960278,0.477133,0.522867,53.953159,23.125414
3,stage4i__candidate_01_base_raw,boosting_base_v1,raw,6,474,8.081507,67.369324,18251.175310,135.096911,44.864300,...,0.009600,67369.324293,135096.910807,True,201.572016,287.862450,0.471533,0.528467,58.109491,15.935148


,experiment_id,feature_pack,target_mode,max_depth,best_iteration,fit_seconds,mae,mse,rmse,mape_percent,...,negative_prediction_rate,mae_usd,rmse_usd,original_scale,top_decile_mae,top_five_percent_mae,underestimation_rate,overestimation_rate,training_mae,validation_train_mae_gap_percent
0,stage4i__candidate_03_sparse_log1p,xgboost_sparse_v1,log1p,6,720,18.142525,64.685719,19237.585130,138.699622,33.859754,...,0.000000,64685.719242,138699.621953,True,214.398433,306.327287,0.540400,0.459600,55.623919,16.291193
1,stage4i__candidate_04_safer_depth5_log1p,xgboost_sparse_v1,log1p,5,1132,26.380924,64.793724,19005.618345,137.860866,33.925681,...,0.000000,64793.723688,137860.865894,True,214.307146,305.693304,0.539200,0.460800,56.445130,14.790636
2,stage4i__candidate_02_sparse_raw,xgboost_sparse_v1,raw,6,688,17.137593,66.430050,18523.315288,136.100387,43.024905,...,0.007267,66430.050314,136100.386804,True,202.129636,287.960278,0.477133,0.522867,53.953159,23.125414


### 9. Optional Refinement

Why: Show the one depth-only refinement.

The code below reads compact saved evidence.

Interpretation: Depth 5 slightly worsened MAE and needed 1,132 trees.

Conclusion: The one allowed refinement was rejected.

In [10]:
display(candidates[candidates.experiment_id.str.contains('candidate_04')])

,experiment_id,feature_pack,target_mode,max_depth,best_iteration,fit_seconds,mae,mse,rmse,mape_percent,...,negative_prediction_rate,mae_usd,rmse_usd,original_scale,top_decile_mae,top_five_percent_mae,underestimation_rate,overestimation_rate,training_mae,validation_train_mae_gap_percent
1,stage4i__candidate_04_safer_depth5_log1p,xgboost_sparse_v1,log1p,5,1132,26.380924,64.793724,19005.618345,137.860866,33.925681,...,0.0,64793.723688,137860.865894,True,214.307146,305.693304,0.5392,0.4608,56.44513,14.790636


### 10. Candidate Results

Why: Compare the bounded results on original target scale.

The code below reads compact saved evidence.

Interpretation: Sparse/log1p had the best MAE at 64.6857; sparse/raw had better upper-tail MAE.

Conclusion: Sparse/log1p was the preliminary direction under the stated selection order.

In [11]:
display(candidates[['experiment_id','feature_pack','target_mode','mae','rmse','rmsle_clipped_zero','top_decile_mae','best_iteration','fit_seconds']])

,experiment_id,feature_pack,target_mode,mae,rmse,rmsle_clipped_zero,top_decile_mae,best_iteration,fit_seconds
0,stage4i__candidate_03_sparse_log1p,xgboost_sparse_v1,log1p,64.685719,138.699622,0.394762,214.398433,720,18.142525
1,stage4i__candidate_04_safer_depth5_log1p,xgboost_sparse_v1,log1p,64.793724,137.860866,0.394813,214.307146,1132,26.380924
2,stage4i__candidate_02_sparse_raw,xgboost_sparse_v1,raw,66.430050,136.100387,0.496130,202.129636,688,17.137593
3,stage4i__candidate_01_base_raw,boosting_base_v1,raw,67.369324,135.096911,0.510010,201.572016,474,8.081507


### 11. Preliminary Configuration

Why: Show the frozen preliminary choice.

The code below reads compact saved evidence.

Interpretation: The preliminary model uses the sparse pack, log1p target, depth 6, and 720 effective trees.

Conclusion: The Stage 4I configuration is frozen and reproducible.

In [12]:
show_json('artifacts/results/stage4/xgboost/initial/stage4i_preliminary_configuration.json')

{'stage': 'stage4i',
 'selected_experiment_id': 'stage4i__candidate_03_sparse_log1p',
 'selection_reason': 'Lowest MAE; within the refinement tie band but uses fewer effective trees and less runtime than depth 5.',
 'feature_pack': 'xgboost_sparse_v1',
 'target_mode': 'log1p',
 'parameters': {'objective': 'reg:squarederror',
  'tree_method': 'hist',
  'device': 'cpu',
  'n_estimators': 1500,
  'learning_rate': 0.05,
  'max_depth': 6,
  'min_child_weight': 10,
  'subsample': 0.8,
  'colsample_bytree': 0.8,
  'reg_alpha': 0.0,
  'reg_lambda': 5.0,
  'gamma': 0.0,
  'max_bin': 256,
  'random_state': 42,
  'n_jobs': 2,
  'eval_metric': 'mae'},
 'best_iteration': 720,
 'metrics': {'mae': 64.68571924163994,
  'mse': 19237.585129877058,
  'rmse': 138.69962195289884,
  'mape_percent': 33.859753568944114,
  'r_squared': 0.6623871005793287,
  'rmsle': 0.39476169368127717,
  'rmsle_clipped_zero': 0.39476169368127717,
  'median_absolute_error': 37.72934561770978,
  'wape_percent': 26.2351717228294

{'stage': 'stage4i',
 'selected_experiment_id': 'stage4i__candidate_03_sparse_log1p',
 'selection_reason': 'Lowest MAE; within the refinement tie band but uses fewer effective trees and less runtime than depth 5.',
 'feature_pack': 'xgboost_sparse_v1',
 'target_mode': 'log1p',
 'parameters': {'objective': 'reg:squarederror',
  'tree_method': 'hist',
  'device': 'cpu',
  'n_estimators': 1500,
  'learning_rate': 0.05,
  'max_depth': 6,
  'min_child_weight': 10,
  'subsample': 0.8,
  'colsample_bytree': 0.8,
  'reg_alpha': 0.0,
  'reg_lambda': 5.0,
  'gamma': 0.0,
  'max_bin': 256,
  'random_state': 42,
  'n_jobs': 2,
  'eval_metric': 'mae'},
 'best_iteration': 720,
 'metrics': {'mae': 64.68571924163994,
  'mse': 19237.585129877058,
  'rmse': 138.69962195289884,
  'mape_percent': 33.859753568944114,
  'r_squared': 0.6623871005793287,
  'rmsle': 0.39476169368127717,
  'rmsle_clipped_zero': 0.39476169368127717,
  'median_absolute_error': 37.72934561770978,
  'wape_percent': 26.2351717228294

### 12. Preliminary Model Saving

Why: Show the saved bundle, native model, and hashes.

The code below reads compact saved evidence.

Interpretation: The promoted bundle, native JSON, predictions, and hashes are recorded.

Conclusion: Only the selected Candidate was promoted as preliminary.

In [13]:
show_json('artifacts/manifests/stage4/xgboost/stage4i_preliminary_model_manifest.json')

{'stage': 'stage4i',
 'selected_experiment_id': 'stage4i__candidate_03_sparse_log1p',
 'selection_reason': 'Lowest MAE; within the refinement tie band but uses fewer effective trees and less runtime than depth 5.',
 'feature_pack': 'xgboost_sparse_v1',
 'target_mode': 'log1p',
 'parameters': {'objective': 'reg:squarederror',
  'tree_method': 'hist',
  'device': 'cpu',
  'n_estimators': 1500,
  'learning_rate': 0.05,
  'max_depth': 6,
  'min_child_weight': 10,
  'subsample': 0.8,
  'colsample_bytree': 0.8,
  'reg_alpha': 0.0,
  'reg_lambda': 5.0,
  'gamma': 0.0,
  'max_bin': 256,
  'random_state': 42,
  'n_jobs': 2,
  'eval_metric': 'mae'},
 'best_iteration': 720,
 'metrics': {'mae': 64.68571924163994,
  'mse': 19237.585129877058,
  'rmse': 138.69962195289884,
  'mape_percent': 33.859753568944114,
  'r_squared': 0.6623871005793287,
  'rmsle': 0.39476169368127717,
  'rmsle_clipped_zero': 0.39476169368127717,
  'median_absolute_error': 37.72934561770978,
  'wape_percent': 26.2351717228294

{'stage': 'stage4i',
 'selected_experiment_id': 'stage4i__candidate_03_sparse_log1p',
 'selection_reason': 'Lowest MAE; within the refinement tie band but uses fewer effective trees and less runtime than depth 5.',
 'feature_pack': 'xgboost_sparse_v1',
 'target_mode': 'log1p',
 'parameters': {'objective': 'reg:squarederror',
  'tree_method': 'hist',
  'device': 'cpu',
  'n_estimators': 1500,
  'learning_rate': 0.05,
  'max_depth': 6,
  'min_child_weight': 10,
  'subsample': 0.8,
  'colsample_bytree': 0.8,
  'reg_alpha': 0.0,
  'reg_lambda': 5.0,
  'gamma': 0.0,
  'max_bin': 256,
  'random_state': 42,
  'n_jobs': 2,
  'eval_metric': 'mae'},
 'best_iteration': 720,
 'metrics': {'mae': 64.68571924163994,
  'mse': 19237.585129877058,
  'rmse': 138.69962195289884,
  'mape_percent': 33.859753568944114,
  'r_squared': 0.6623871005793287,
  'rmsle': 0.39476169368127717,
  'rmsle_clipped_zero': 0.39476169368127717,
  'median_absolute_error': 37.72934561770978,
  'wape_percent': 26.2351717228294

### 13. Clean-Process Reload

Why: Confirm raw-frame and native prediction matching.

The code below reads compact saved evidence.

Interpretation: Bundle, native, and saved reference predictions match on 500 raw rows; the matrix has 158 sparse columns.

Conclusion: Preliminary serialization is valid.

In [14]:
show_json('artifacts/reports/stage4i_preliminary_reload.json')

{'model_path': 'artifacts\\models\\xgboost\\preliminary\\xgboost_preliminary_without_sensitive.joblib',
 'native_model_path': 'artifacts\\models\\xgboost\\preliminary\\xgboost_preliminary_without_sensitive.json',
 'prediction_path': 'artifacts\\predictions\\xgboost\\initial\\xgboost_preliminary_without_sensitive.csv',
 'rows': 500,
 'finite_predictions': True,
 'output_length_matches': True,
 'source_frame_unchanged': True,
 'reference_prediction_match': True,
 'native_prediction_match': True,
 'inverse_target_verified': True,
 'native_model_loaded': True,
 'sparse_transformation': True,
 'sparse_shape': [500, 158],
 'status': 'PASS'}

{'model_path': 'artifacts\\models\\xgboost\\preliminary\\xgboost_preliminary_without_sensitive.joblib',
 'native_model_path': 'artifacts\\models\\xgboost\\preliminary\\xgboost_preliminary_without_sensitive.json',
 'prediction_path': 'artifacts\\predictions\\xgboost\\initial\\xgboost_preliminary_without_sensitive.csv',
 'rows': 500,
 'finite_predictions': True,
 'output_length_matches': True,
 'source_frame_unchanged': True,
 'reference_prediction_match': True,
 'native_prediction_match': True,
 'inverse_target_verified': True,
 'native_model_loaded': True,
 'sparse_transformation': True,
 'sparse_shape': [500, 158],
 'status': 'PASS'}

### 14. Gain and Weight Importance

Why: Show model importance without a causal claim.

The code below reads compact saved evidence.

Interpretation: Lien status, applicant income, estimated tract income, and lien interactions lead total gain.

Conclusion: Importance supports focused Feature proposals but is not causal.

In [15]:
show_csv('artifacts/results/stage4/xgboost/initial/stage4i_importance_top20.csv',20)

,source_feature,gain,weight,total_gain
0,lien_status_name,1876.993294,169.0,124850.857910
1,applicant_income_000s,29.339890,2125.0,62347.265625
2,estimated_tract_family_income_000s,9.542030,1802.0,17194.738281
3,loan_type_lien_status_group,1533.109894,216.0,11014.403219
4,applicant_income_to_area_income,4.923801,1722.0,8478.786133
5,us_region,48.537259,488.0,7344.291679
6,applicant_vs_area_income_gap_000s,4.708585,1414.0,6657.938965
7,hud_median_family_income,4.202759,1573.0,6610.940430
8,state_name,100.095943,1459.0,5204.120084
9,state_lien_status_group,4.300886,1171.0,5036.337891


,source_feature,gain,weight,total_gain
0,lien_status_name,1876.993294,169.0,124850.857910
1,applicant_income_000s,29.339890,2125.0,62347.265625
2,estimated_tract_family_income_000s,9.542030,1802.0,17194.738281
3,loan_type_lien_status_group,1533.109894,216.0,11014.403219
4,applicant_income_to_area_income,4.923801,1722.0,8478.786133
5,us_region,48.537259,488.0,7344.291679
6,applicant_vs_area_income_gap_000s,4.708585,1414.0,6657.938965
7,hud_median_family_income,4.202759,1573.0,6610.940430
8,state_name,100.095943,1459.0,5204.120084
9,state_lien_status_group,4.300886,1171.0,5036.337891


### 15. Bounded Native SHAP

Why: Show bounded native contribution evidence.

The code below reads compact saved evidence.

Interpretation: Native SHAP uses 300 fixed Validation IDs and also highlights lien and income signals.

Conclusion: The bounded interpretation agrees with Importance at a broad level.

In [16]:
show_csv('artifacts/results/stage4/xgboost/initial/stage4i_shap_top20.csv',20); show_json('artifacts/manifests/stage4/xgboost/stage4i_shap_metadata.json')

,source_feature,mean_absolute_shap
0,lien_status_name,0.251966
1,applicant_income_000s,0.224113
2,estimated_tract_family_income_000s,0.109567
3,us_region,0.101304
4,property_purpose_group,0.086295
5,state_name,0.058667
6,hud_median_family_income,0.056585
7,loan_type_name,0.042066
8,occupancy_lien_status_group,0.041453
9,state_lien_status_group,0.038347


{'stage': 'stage4i',
 'model_path': 'artifacts\\models\\xgboost\\preliminary\\xgboost_preliminary_without_sensitive.joblib',
 'rows': 300,
 'row_ids_path': 'artifacts/manifests/stage4/xgboost/stage4i_shap_row_ids.csv',
 'target_output_scale': 'log1p model output',
 'native_pred_contribs': True,
 'kernel_shap_used': False,
 'status': 'PASS'}

{'stage': 'stage4i',
 'model_path': 'artifacts\\models\\xgboost\\preliminary\\xgboost_preliminary_without_sensitive.joblib',
 'rows': 300,
 'row_ids_path': 'artifacts/manifests/stage4/xgboost/stage4i_shap_row_ids.csv',
 'target_output_scale': 'log1p model output',
 'native_pred_contribs': True,
 'kernel_shap_used': False,
 'status': 'PASS'}

### 16. Error Analysis

Why: Show tail error and signed error patterns.

The code below reads compact saved evidence.

Interpretation: Top-decile MAE is 214.40 and mean signed error is -16.76, showing material upper-tail underestimation.

Conclusion: Tail error must remain visible in later selection.

In [17]:
show_csv('artifacts/results/stage4/xgboost/initial/stage4i_error_by_target_decile.csv',10); show_json('artifacts/results/stage4/xgboost/initial/stage4i_error_summary.json')

,target_decile,rows,target_min,target_max,mae,mean_signed_error,underestimation_rate,overestimation_rate
0,0,1502,1.0,64.0,25.327635,20.645447,0.266312,0.733688
1,1,1517,65.0,106.0,39.534131,30.494015,0.246539,0.753461
2,2,1493,107.0,137.0,38.310182,21.792544,0.367716,0.632284
3,3,1515,138.0,166.0,36.411232,12.223355,0.454785,0.545215
4,4,1485,167.0,197.0,37.783089,3.519565,0.528620,0.471380
5,5,1491,198.0,232.0,45.750223,-0.931261,0.580148,0.419852
6,6,1503,233.0,277.0,51.693951,-17.813939,0.677977,0.322023
7,7,1513,278.0,344.0,67.421565,-29.850494,0.721745,0.278255
8,8,1483,345.0,444.0,90.196865,-48.343731,0.743088,0.256912
9,9,1498,445.0,8075.0,215.034732,-160.153061,0.821762,0.178238


{'top_decile_mae': 214.39843309134062,
 'top_five_percent_mae': 306.3272871161781,
 'underestimation_rate': 0.5404,
 'overestimation_rate': 0.4596,
 'mean_signed_error': -16.75864044745151,
 'largest_decile_mae': 215.03473236662916,
 'largest_decile': 9}

{'top_decile_mae': 214.39843309134062,
 'top_five_percent_mae': 306.3272871161781,
 'underestimation_rate': 0.5404,
 'overestimation_rate': 0.4596,
 'mean_signed_error': -16.75864044745151,
 'largest_decile_mae': 215.03473236662916,
 'largest_decile': 9}

### 17. Stage 4J Feature Proposals

Why: Show no more than three safe proposals.

The code below reads compact saved evidence.

Interpretation: Three proposals use only non-sensitive Feature values and no target encoding.

Conclusion: They are proposals only until independent confirmation.

In [18]:
show_csv('artifacts/features/stage4/xgboost/stage4i_feature_proposals.csv',3)

,name,formula,source_columns,evidence,expected_benefit,fixed_or_learned,missing_handling,zero_denominator_handling,target_derived,sensitive_derived,leakage_review,duplicate_review
0,applicant_to_estimated_tract_income_ratio,applicant_income_000s / estimated_tract_family...,applicant_income_000s; hud_median_family_incom...,income gain rank=2; income SHAP rank=2; top-de...,Represent applicant income relative to the est...,fixed row formula,missing source gives missing output,epsilon denominator plus explicit zero flag,False,False,PASS: raw non-sensitive row fields only,Not an exact existing Feature; existing income...
1,respondent_purpose_group,"respondent_id + loan_purpose_name, then traini...",respondent_id; loan_purpose_name,respondent frequency gain rank=16; loan purpos...,Capture lender-purpose patterns without target...,fixed interaction plus training-fit frequency map,stable missing token,not applicable,False,False,PASS: mapping uses Feature frequency on traini...,No existing respondent-purpose interaction.
2,income_band_lien_status_group,training-fit applicant-income quantile band + ...,applicant_income_000s; lien_status_name,income SHAP rank=2; lien gain rank=1; upper-ta...,Allow a compact non-linear income and lien int...,training-fit income bins plus fixed category i...,stable missing token,not applicable,False,False,PASS: quantiles fit on training Feature values...,No existing income-band and lien interaction.


,name,formula,source_columns,evidence,expected_benefit,fixed_or_learned,missing_handling,zero_denominator_handling,target_derived,sensitive_derived,leakage_review,duplicate_review
0,applicant_to_estimated_tract_income_ratio,applicant_income_000s / estimated_tract_family...,applicant_income_000s; hud_median_family_incom...,income gain rank=2; income SHAP rank=2; top-de...,Represent applicant income relative to the est...,fixed row formula,missing source gives missing output,epsilon denominator plus explicit zero flag,False,False,PASS: raw non-sensitive row fields only,Not an exact existing Feature; existing income...
1,respondent_purpose_group,"respondent_id + loan_purpose_name, then traini...",respondent_id; loan_purpose_name,respondent frequency gain rank=16; loan purpos...,Capture lender-purpose patterns without target...,fixed interaction plus training-fit frequency map,stable missing token,not applicable,False,False,PASS: mapping uses Feature frequency on traini...,No existing respondent-purpose interaction.
2,income_band_lien_status_group,training-fit applicant-income quantile band + ...,applicant_income_000s; lien_status_name,income SHAP rank=2; lien gain rank=1; upper-ta...,Allow a compact non-linear income and lien int...,training-fit income bins plus fixed category i...,stable missing token,not applicable,False,False,PASS: quantiles fit on training Feature values...,No existing income-band and lien interaction.


### 18. Stage 4I Gate Summary

Why: Confirm that Gate 1 passed before Stage 4J.

The code below reads compact saved evidence.

Interpretation: Every Stage 4I Gate check is true, including zero Test overlap and unique Registry IDs.

Conclusion: Stage 4J was allowed to start.

In [19]:
show_json('artifacts/reports/stage4i_gate_verification.json')

{'stage': 'stage4i',
 'official_name': 'Stage 4I — Initial XGBoost Model and Importance Analysis',
 'created_at_utc': '2026-07-14T19:16:32.852882+00:00',
 'candidate_count': 4,
 'required_candidate_count': 3,
 'optional_refinement_count': 1,
 'selected_experiment_id': 'stage4i__candidate_03_sparse_log1p',
 'prediction_test_overlaps': {'artifacts\\predictions\\xgboost\\initial\\candidate_01_base_raw.csv': 0,
  'artifacts\\predictions\\xgboost\\initial\\candidate_02_sparse_raw.csv': 0,
  'artifacts\\predictions\\xgboost\\initial\\candidate_03_sparse_log1p.csv': 0,
  'artifacts\\predictions\\xgboost\\initial\\candidate_04_safer_depth5_log1p.csv': 0,
  'artifacts\\predictions\\xgboost\\initial\\xgboost_preliminary_without_sensitive.csv': 0},
 'checks': {'discovery_sample_valid': True,
  'test_overlap_zero': True,
  'xgboost_import_and_cpu_mode_valid': True,
  'three_required_candidates_complete': True,
  'total_candidate_budget_at_most_four': True,
  'raw_and_log1p_evaluated': True,
  'all

{'stage': 'stage4i',
 'official_name': 'Stage 4I — Initial XGBoost Model and Importance Analysis',
 'created_at_utc': '2026-07-14T19:16:32.852882+00:00',
 'candidate_count': 4,
 'required_candidate_count': 3,
 'optional_refinement_count': 1,
 'selected_experiment_id': 'stage4i__candidate_03_sparse_log1p',
 'prediction_test_overlaps': {'artifacts\\predictions\\xgboost\\initial\\candidate_01_base_raw.csv': 0,
  'artifacts\\predictions\\xgboost\\initial\\candidate_02_sparse_raw.csv': 0,
  'artifacts\\predictions\\xgboost\\initial\\candidate_03_sparse_log1p.csv': 0,
  'artifacts\\predictions\\xgboost\\initial\\candidate_04_safer_depth5_log1p.csv': 0,
  'artifacts\\predictions\\xgboost\\initial\\xgboost_preliminary_without_sensitive.csv': 0},
 'checks': {'discovery_sample_valid': True,
  'test_overlap_zero': True,
  'xgboost_import_and_cpu_mode_valid': True,
  'three_required_candidates_complete': True,
  'total_candidate_budget_at_most_four': True,
  'raw_and_log1p_evaluated': True,
  'all

## Stage 4J — XGBoost Feature Confirmation and Final Tuning

### 19. Stage 4J Objective

Why: State the independent confirmation and tuning scope.

The code below reads compact saved evidence.

Interpretation: Stage 4J separates Feature confirmation from final tuning and remains non-sensitive.

Conclusion: Sensitive data cannot influence the final pack or parameters.

In [20]:
print('Stage 4J uses non-sensitive Feature Confirmation and Final Selection rows only.')

Stage 4J uses non-sensitive Feature Confirmation and Final Selection rows only.


### 20. Stage 4I Gate Validation

Why: Recheck the Gate dependency.

The code below reads compact saved evidence.

Interpretation: The Stage 4I Gate remains PASS.

Conclusion: Feature confirmation rests on completed preliminary evidence.

In [21]:
show_json('artifacts/reports/stage4i_gate_verification.json')

{'stage': 'stage4i',
 'official_name': 'Stage 4I — Initial XGBoost Model and Importance Analysis',
 'created_at_utc': '2026-07-14T19:16:32.852882+00:00',
 'candidate_count': 4,
 'required_candidate_count': 3,
 'optional_refinement_count': 1,
 'selected_experiment_id': 'stage4i__candidate_03_sparse_log1p',
 'prediction_test_overlaps': {'artifacts\\predictions\\xgboost\\initial\\candidate_01_base_raw.csv': 0,
  'artifacts\\predictions\\xgboost\\initial\\candidate_02_sparse_raw.csv': 0,
  'artifacts\\predictions\\xgboost\\initial\\candidate_03_sparse_log1p.csv': 0,
  'artifacts\\predictions\\xgboost\\initial\\candidate_04_safer_depth5_log1p.csv': 0,
  'artifacts\\predictions\\xgboost\\initial\\xgboost_preliminary_without_sensitive.csv': 0},
 'checks': {'discovery_sample_valid': True,
  'test_overlap_zero': True,
  'xgboost_import_and_cpu_mode_valid': True,
  'three_required_candidates_complete': True,
  'total_candidate_budget_at_most_four': True,
  'raw_and_log1p_evaluated': True,
  'all

{'stage': 'stage4i',
 'official_name': 'Stage 4I — Initial XGBoost Model and Importance Analysis',
 'created_at_utc': '2026-07-14T19:16:32.852882+00:00',
 'candidate_count': 4,
 'required_candidate_count': 3,
 'optional_refinement_count': 1,
 'selected_experiment_id': 'stage4i__candidate_03_sparse_log1p',
 'prediction_test_overlaps': {'artifacts\\predictions\\xgboost\\initial\\candidate_01_base_raw.csv': 0,
  'artifacts\\predictions\\xgboost\\initial\\candidate_02_sparse_raw.csv': 0,
  'artifacts\\predictions\\xgboost\\initial\\candidate_03_sparse_log1p.csv': 0,
  'artifacts\\predictions\\xgboost\\initial\\candidate_04_safer_depth5_log1p.csv': 0,
  'artifacts\\predictions\\xgboost\\initial\\xgboost_preliminary_without_sensitive.csv': 0},
 'checks': {'discovery_sample_valid': True,
  'test_overlap_zero': True,
  'xgboost_import_and_cpu_mode_valid': True,
  'three_required_candidates_complete': True,
  'total_candidate_budget_at_most_four': True,
  'raw_and_log1p_evaluated': True,
  'all

### 21. Feature Proposal Review

Why: Show accepted and rejected proposals.

The code below reads compact saved evidence.

Interpretation: Only the income-to-estimated-tract-income ratio was accepted; two interactions were rejected.

Conclusion: The final pack did not force all proposals to succeed.

In [22]:
show_csv('artifacts/features/stage4/xgboost/stage4j_feature_decisions.csv',3)

,proposal,decision,rule,reason
0,applicant_to_estimated_tract_income_ratio,accepted,stability,"MAE, RMSE, P90, signed bias, top-decile and to..."
1,respondent_purpose_group,rejected,combined pack failed,No independent acceptance evidence; combined p...
2,income_band_lien_status_group,rejected,combined pack failed,No independent acceptance evidence; combined p...


,proposal,decision,rule,reason
0,applicant_to_estimated_tract_income_ratio,accepted,stability,"MAE, RMSE, P90, signed bias, top-decile and to..."
1,respondent_purpose_group,rejected,combined pack failed,No independent acceptance evidence; combined p...
2,income_band_lien_status_group,rejected,combined pack failed,No independent acceptance evidence; combined p...


### 22. XGBoost Feature Engineer v2

Why: Confirm stable, serializable Feature engineering.

The code below reads compact saved evidence.

Interpretation: Feature Engineer v2 preserves rows and columns, does not mutate input, and reloads in a clean process.

Conclusion: The accepted Feature can live inside saved bundles.

In [23]:
show_json('artifacts/reports/stage4j_feature_engineer_v2_roundtrip.json')

{'stage': 'stage4j',
 'rows': 1000,
 'selected_proposals': ['applicant_to_estimated_tract_income_ratio',
  'respondent_purpose_group',
  'income_band_lien_status_group'],
 'source_unchanged': True,
 'row_order_preserved': True,
 'stable_columns': True,
 'joblib_roundtrip_match': True,
 'clean_process_import_transform': True,
 'clean_process_stdout': '20 39',
 'target_used': False,
 'sensitive_sources_used': False,
 'status': 'PASS'}

{'stage': 'stage4j',
 'rows': 1000,
 'selected_proposals': ['applicant_to_estimated_tract_income_ratio',
  'respondent_purpose_group',
  'income_band_lien_status_group'],
 'source_unchanged': True,
 'row_order_preserved': True,
 'stable_columns': True,
 'joblib_roundtrip_match': True,
 'clean_process_import_transform': True,
 'clean_process_stdout': '20 39',
 'target_used': False,
 'sensitive_sources_used': False,
 'status': 'PASS'}

### 23. Feature Confirmation Sample

Why: Show the independent 80,000/20,000 roles.

The code below reads compact saved evidence.

Interpretation: Feature Confirmation has independent 80,000 Train and 20,000 Validation rows.

Conclusion: Proposal decisions do not reuse Discovery Validation evidence.

In [24]:
display(samples['samples']['feature_confirmation'])

{'rows': 100000,
 'train_rows': 80000,
 'validation_rows': 20000,
 'expected_rows': 100000,
 'row_ids_unique': True,
 'maximum_target_bin_proportion_difference': 1.2069997098462792e-05,
 'role_maximum_target_bin_proportion_difference': {'train': 7.06999709847167e-06,
  'validation': 3.2069997098468916e-05},
 'sha256': '2dabbdc5e6b0fee9bc50e63feeb912e3096a5408cafbbf1685c1e00332380cf4',
 'valid': True}

### 24. Lean Feature Confirmation

Why: Show the original, combined, and ratio-only fits.

The code below reads compact saved evidence.

Interpretation: The ratio-only pack improved MAE by 0.275%, RMSE by 2.99%, and several tail/bias metrics with bounded runtime cost.

Conclusion: It passed the stability rule; the combined pack did not.

In [25]:
show_csv('artifacts/results/stage4/xgboost/feature_confirmation/stage4j_feature_confirmation_results.csv',3)

,experiment_id,feature_pack,selected_proposals,best_iteration,fit_seconds,mae,mse,rmse,mape_percent,r_squared,...,negative_prediction_rate,mae_usd,rmse_usd,original_scale,top_decile_mae,top_five_percent_mae,underestimation_rate,overestimation_rate,mae_improvement_percent_vs_original,top_decile_improvement_percent_vs_original
0,stage4j__confirmation_original,xgboost_sparse_v1,NaN,877,29.445286,64.827476,17277.237664,131.442906,33.666982,0.699225,...,0.0,64827.476077,131442.906481,True,213.608413,308.273779,0.53520,0.46480,0.000000,0.000000
1,stage4j__confirmation_combined_v2,xgboost_sparse_v2_all,applicant_to_estimated_tract_income_ratio;resp...,1136,41.805985,64.660601,16350.189285,127.867859,33.470518,0.715364,...,0.0,64660.601469,127867.858685,True,212.070904,304.310205,0.53785,0.46215,0.257413,0.719779
2,stage4j__confirmation_ratio_only_v2,xgboost_sparse_v2_ratio,applicant_to_estimated_tract_income_ratio,1032,34.905303,64.649102,16260.437707,127.516421,33.570248,0.716927,...,0.0,64649.102372,127516.421323,True,212.272901,304.052661,0.53785,0.46215,0.275151,0.625215


,experiment_id,feature_pack,selected_proposals,best_iteration,fit_seconds,mae,mse,rmse,mape_percent,r_squared,...,negative_prediction_rate,mae_usd,rmse_usd,original_scale,top_decile_mae,top_five_percent_mae,underestimation_rate,overestimation_rate,mae_improvement_percent_vs_original,top_decile_improvement_percent_vs_original
0,stage4j__confirmation_original,xgboost_sparse_v1,NaN,877,29.445286,64.827476,17277.237664,131.442906,33.666982,0.699225,...,0.0,64827.476077,131442.906481,True,213.608413,308.273779,0.53520,0.46480,0.000000,0.000000
1,stage4j__confirmation_combined_v2,xgboost_sparse_v2_all,applicant_to_estimated_tract_income_ratio;resp...,1136,41.805985,64.660601,16350.189285,127.867859,33.470518,0.715364,...,0.0,64660.601469,127867.858685,True,212.070904,304.310205,0.53785,0.46215,0.257413,0.719779
2,stage4j__confirmation_ratio_only_v2,xgboost_sparse_v2_ratio,applicant_to_estimated_tract_income_ratio,1032,34.905303,64.649102,16260.437707,127.516421,33.570248,0.716927,...,0.0,64649.102372,127516.421323,True,212.272901,304.052661,0.53785,0.46215,0.275151,0.625215


### 25. Final XGBoost Feature Pack

Why: Show the honestly frozen ratio-only pack.

The code below reads compact saved evidence.

Interpretation: The frozen pack adds only the accepted ratio and zero-denominator flag to the Stage 4B sparse pack.

Conclusion: Both sensitive modes use the same final Feature policy.

In [26]:
show_json('artifacts/features/stage4/xgboost/stage4j_final_feature_pack.json')

{'stage': 'stage4j',
 'feature_pack': {'numeric': ['applicant_income_000s',
   'population',
   'hud_median_family_income',
   'number_of_owner_occupied_units',
   'number_of_1_to_4_family_units',
   'applicant_income_to_area_income',
   'tract_income_ratio',
   'owner_occupied_unit_ratio',
   'family_units_per_1000_people',
   'owner_occupied_units_per_1000_people',
   'has_co_applicant',
   'estimated_tract_family_income_000s',
   'applicant_vs_area_income_gap_000s',
   'applicant_to_estimated_tract_income_ratio',
   'applicant_to_estimated_tract_income_denominator_zero'],
  'one_hot_categorical': ['agency_name',
   'loan_type_name',
   'property_type_name',
   'loan_purpose_name',
   'owner_occupancy_name',
   'preapproval_name',
   'state_name',
   'lien_status_name',
   'loan_program_group',
   'applicant_income_area_group',
   'tract_income_level',
   'us_region',
   'purpose_lien_status_group',
   'occupancy_lien_status_group',
   'loan_type_lien_status_group',
   'property_purp

{'stage': 'stage4j',
 'feature_pack': {'numeric': ['applicant_income_000s',
   'population',
   'hud_median_family_income',
   'number_of_owner_occupied_units',
   'number_of_1_to_4_family_units',
   'applicant_income_to_area_income',
   'tract_income_ratio',
   'owner_occupied_unit_ratio',
   'family_units_per_1000_people',
   'owner_occupied_units_per_1000_people',
   'has_co_applicant',
   'estimated_tract_family_income_000s',
   'applicant_vs_area_income_gap_000s',
   'applicant_to_estimated_tract_income_ratio',
   'applicant_to_estimated_tract_income_denominator_zero'],
  'one_hot_categorical': ['agency_name',
   'loan_type_name',
   'property_type_name',
   'loan_purpose_name',
   'owner_occupancy_name',
   'preapproval_name',
   'state_name',
   'lien_status_name',
   'loan_program_group',
   'applicant_income_area_group',
   'tract_income_level',
   'us_region',
   'purpose_lien_status_group',
   'occupancy_lien_status_group',
   'loan_type_lien_status_group',
   'property_purp

### 26. Final Selection Sample

Why: Show the independent 100,000/25,000 roles.

The code below reads compact saved evidence.

Interpretation: Final Selection has independent 100,000 Train and 25,000 Validation rows.

Conclusion: The tuning decision uses its designated sample only.

In [27]:
display(samples['samples']['final_selection'])

{'rows': 125000,
 'train_rows': 100000,
 'validation_rows': 25000,
 'expected_rows': 125000,
 'row_ids_unique': True,
 'maximum_target_bin_proportion_difference': 7.974866679333337e-06,
 'role_maximum_target_bin_proportion_difference': {'train': 5.974866679345214e-06,
  'validation': 1.5974866679341337e-05},
 'sha256': '445a077b3fc1963bec38d8ba90384df2e1400e3897b25f0ecbfce78cb96c49ef',
 'valid': True}

### 27. Four Final Tuning Candidates

Why: Show exactly four predeclared Candidates.

The code below reads compact saved evidence.

Interpretation: All four Candidates completed; low-rate D had MAE 64.1664 and A had MAE 64.2219.

Conclusion: The required low-learning-rate direction was evaluated.

In [28]:
tuning=show_csv('artifacts/results/stage4/xgboost/final/stage4j_final_tuning_results.csv',4); display(tuning[['experiment_id','mae','rmse','rmsle','top_decile_mae','best_iteration','fit_seconds']])

,experiment_id,feature_pack,target_mode,max_depth,learning_rate,n_estimators_max,best_iteration,fit_seconds,mae,mse,...,mean_signed_error,p90_absolute_error,negative_prediction_rate,mae_usd,rmse_usd,original_scale,top_decile_mae,top_five_percent_mae,underestimation_rate,overestimation_rate
0,stage4j__tuning_d_low_lr,xgboost_sparse_v2_ratio,log1p,6,0.03,2500,1951,80.266765,64.166380,17029.301349,...,-15.709380,135.944053,0.0,64166.380167,130496.365271,True,209.361848,302.278963,0.53756,0.46244
1,stage4j__tuning_a_baseline,xgboost_sparse_v2_ratio,log1p,6,0.05,1500,982,42.563350,64.221924,17067.861411,...,-15.976944,136.817039,0.0,64221.923647,130644.025547,True,209.422317,302.260959,0.53736,0.46264
2,stage4j__tuning_c_flexible,xgboost_sparse_v2_ratio,log1p,8,0.05,1500,760,38.680711,64.255032,17198.022981,...,-15.668741,136.929740,0.0,64255.031587,131141.232954,True,209.961148,302.450701,0.53716,0.46284
3,stage4j__tuning_b_safer,xgboost_sparse_v2_ratio,log1p,5,0.05,1500,1500,56.576064,64.430620,17276.655548,...,-15.812408,136.762059,0.0,64430.620430,131440.692132,True,210.089300,303.304861,0.53848,0.46152


,experiment_id,mae,rmse,rmsle,top_decile_mae,best_iteration,fit_seconds
0,stage4j__tuning_d_low_lr,64.166380,130.496365,0.390539,209.361848,1951,80.266765
1,stage4j__tuning_a_baseline,64.221924,130.644026,0.391395,209.422317,982,42.563350
2,stage4j__tuning_c_flexible,64.255032,131.141233,0.391762,209.961148,760,38.680711
3,stage4j__tuning_b_safer,64.430620,131.440692,0.391868,210.089300,1500,56.576064


### 28. Final Non-Sensitive Configuration

Why: Show the tie-rule winner and fixed tree count.

The code below reads compact saved evidence.

Interpretation: A won the 0.25% tie rule with 982 trees and about half D?s runtime.

Conclusion: Depth 6, learning rate 0.05, log1p, seed 42, CPU, and two threads are frozen.

In [29]:
show_json('artifacts/results/stage4/xgboost/final/stage4j_final_configuration.json')

{'stage': 'stage4j',
 'selected_experiment_id': 'stage4j__tuning_a_baseline',
 'selection_reason': 'Candidate D had 0.086 percent lower MAE, inside the 0.25 percent tie band; Candidate A uses about half as many trees and runtime.',
 'feature_pack': 'xgboost_sparse_v2_ratio',
 'feature_engineer_version': 'xgboost_feature_engineer_v2_20260714',
 'selected_proposals': ['applicant_to_estimated_tract_income_ratio'],
 'target_mode': 'log1p',
 'parameters': {'objective': 'reg:squarederror',
  'tree_method': 'hist',
  'device': 'cpu',
  'n_estimators': 982,
  'learning_rate': 0.05,
  'max_depth': 6,
  'min_child_weight': 10,
  'subsample': 0.8,
  'colsample_bytree': 0.8,
  'reg_alpha': 0.0,
  'reg_lambda': 5.0,
  'gamma': 0.0,
  'max_bin': 256,
  'random_state': 42,
  'n_jobs': 2,
  'eval_metric': 'mae'},
 'best_iteration': 982,
 'random_seed': 42,
 'execution_mode': 'cpu',
 'thread_count': 2,
 'encoding_policy': 'sparse one-hot plus training-fit frequency maps',
 'metrics': {'mae': 64.2219236

{'stage': 'stage4j',
 'selected_experiment_id': 'stage4j__tuning_a_baseline',
 'selection_reason': 'Candidate D had 0.086 percent lower MAE, inside the 0.25 percent tie band; Candidate A uses about half as many trees and runtime.',
 'feature_pack': 'xgboost_sparse_v2_ratio',
 'feature_engineer_version': 'xgboost_feature_engineer_v2_20260714',
 'selected_proposals': ['applicant_to_estimated_tract_income_ratio'],
 'target_mode': 'log1p',
 'parameters': {'objective': 'reg:squarederror',
  'tree_method': 'hist',
  'device': 'cpu',
  'n_estimators': 982,
  'learning_rate': 0.05,
  'max_depth': 6,
  'min_child_weight': 10,
  'subsample': 0.8,
  'colsample_bytree': 0.8,
  'reg_alpha': 0.0,
  'reg_lambda': 5.0,
  'gamma': 0.0,
  'max_bin': 256,
  'random_state': 42,
  'n_jobs': 2,
  'eval_metric': 'mae'},
 'best_iteration': 982,
 'random_seed': 42,
 'execution_mode': 'cpu',
 'thread_count': 2,
 'encoding_policy': 'sparse one-hot plus training-fit frequency maps',
 'metrics': {'mae': 64.2219236

### 29. Winning Non-Sensitive Validation Model

Why: Show the model later reused by Stage 4K.

The code below reads compact saved evidence.

Interpretation: The exact Candidate A bundle, native JSON, predictions, and hashes are saved.

Conclusion: This is the non-sensitive model Stage 4K must reuse.

In [30]:
show_json('artifacts/manifests/stage4/xgboost/stage4j_winning_non_sensitive_manifest.json')

{'stage': 'stage4j',
 'selected_experiment_id': 'stage4j__tuning_a_baseline',
 'selection_reason': 'Candidate D had 0.086 percent lower MAE, inside the 0.25 percent tie band; Candidate A uses about half as many trees and runtime.',
 'feature_pack': 'xgboost_sparse_v2_ratio',
 'feature_engineer_version': 'xgboost_feature_engineer_v2_20260714',
 'selected_proposals': ['applicant_to_estimated_tract_income_ratio'],
 'target_mode': 'log1p',
 'parameters': {'objective': 'reg:squarederror',
  'tree_method': 'hist',
  'device': 'cpu',
  'n_estimators': 982,
  'learning_rate': 0.05,
  'max_depth': 6,
  'min_child_weight': 10,
  'subsample': 0.8,
  'colsample_bytree': 0.8,
  'reg_alpha': 0.0,
  'reg_lambda': 5.0,
  'gamma': 0.0,
  'max_bin': 256,
  'random_state': 42,
  'n_jobs': 2,
  'eval_metric': 'mae'},
 'best_iteration': 982,
 'random_seed': 42,
 'execution_mode': 'cpu',
 'thread_count': 2,
 'encoding_policy': 'sparse one-hot plus training-fit frequency maps',
 'metrics': {'mae': 64.2219236

{'stage': 'stage4j',
 'selected_experiment_id': 'stage4j__tuning_a_baseline',
 'selection_reason': 'Candidate D had 0.086 percent lower MAE, inside the 0.25 percent tie band; Candidate A uses about half as many trees and runtime.',
 'feature_pack': 'xgboost_sparse_v2_ratio',
 'feature_engineer_version': 'xgboost_feature_engineer_v2_20260714',
 'selected_proposals': ['applicant_to_estimated_tract_income_ratio'],
 'target_mode': 'log1p',
 'parameters': {'objective': 'reg:squarederror',
  'tree_method': 'hist',
  'device': 'cpu',
  'n_estimators': 982,
  'learning_rate': 0.05,
  'max_depth': 6,
  'min_child_weight': 10,
  'subsample': 0.8,
  'colsample_bytree': 0.8,
  'reg_alpha': 0.0,
  'reg_lambda': 5.0,
  'gamma': 0.0,
  'max_bin': 256,
  'random_state': 42,
  'n_jobs': 2,
  'eval_metric': 'mae'},
 'best_iteration': 982,
 'random_seed': 42,
 'execution_mode': 'cpu',
 'thread_count': 2,
 'encoding_policy': 'sparse one-hot plus training-fit frequency maps',
 'metrics': {'mae': 64.2219236

### 30. Clean-Process Reload

Why: Confirm winning bundle and native model matching.

The code below reads compact saved evidence.

Interpretation: The winning bundle and native model match the saved predictions on a sparse 159-column matrix.

Conclusion: The Stage 4J winner is safe to reuse.

In [31]:
show_json('artifacts/reports/stage4j_winning_model_reload.json')

{'model_path': 'artifacts\\models\\xgboost\\final\\xgboost_winning_non_sensitive_validation.joblib',
 'native_model_path': 'artifacts\\models\\xgboost\\final\\xgboost_winning_non_sensitive_validation.json',
 'prediction_path': 'artifacts\\predictions\\xgboost\\final_selection\\xgboost_winning_non_sensitive_validation.csv',
 'rows': 500,
 'finite_predictions': True,
 'output_length_matches': True,
 'source_frame_unchanged': True,
 'reference_prediction_match': True,
 'native_prediction_match': True,
 'inverse_target_verified': True,
 'native_model_loaded': True,
 'sparse_transformation': True,
 'sparse_shape': [500, 159],
 'status': 'PASS'}

{'model_path': 'artifacts\\models\\xgboost\\final\\xgboost_winning_non_sensitive_validation.joblib',
 'native_model_path': 'artifacts\\models\\xgboost\\final\\xgboost_winning_non_sensitive_validation.json',
 'prediction_path': 'artifacts\\predictions\\xgboost\\final_selection\\xgboost_winning_non_sensitive_validation.csv',
 'rows': 500,
 'finite_predictions': True,
 'output_length_matches': True,
 'source_frame_unchanged': True,
 'reference_prediction_match': True,
 'native_prediction_match': True,
 'inverse_target_verified': True,
 'native_model_loaded': True,
 'sparse_transformation': True,
 'sparse_shape': [500, 159],
 'status': 'PASS'}

### 31. Stage 4J Gate Summary

Why: Confirm Gate 2 before sensitive fitting.

The code below reads compact saved evidence.

Interpretation: All Stage 4J checks pass, including three confirmation fits and exactly four tuning fits.

Conclusion: Stage 4K was allowed to begin.

In [32]:
show_json('artifacts/reports/stage4j_gate_verification.json')

{'stage': 'stage4j',
 'official_name': 'Stage 4J — XGBoost Feature Confirmation and Final Tuning',
 'created_at_utc': '2026-07-14T19:27:37.857116+00:00',
 'confirmation_fit_count': 3,
 'tuning_fit_count': 4,
 'selected_experiment_id': 'stage4j__tuning_a_baseline',
 'prediction_test_overlaps': {'artifacts\\predictions\\xgboost\\feature_confirmation\\confirmation_combined_v2.csv': 0,
  'artifacts\\predictions\\xgboost\\feature_confirmation\\confirmation_original.csv': 0,
  'artifacts\\predictions\\xgboost\\feature_confirmation\\confirmation_ratio_only_v2.csv': 0,
  'artifacts\\predictions\\xgboost\\final_selection\\tuning_a_baseline.csv': 0,
  'artifacts\\predictions\\xgboost\\final_selection\\tuning_b_safer.csv': 0,
  'artifacts\\predictions\\xgboost\\final_selection\\tuning_c_flexible.csv': 0,
  'artifacts\\predictions\\xgboost\\final_selection\\tuning_d_low_lr.csv': 0,
  'artifacts\\predictions\\xgboost\\final_selection\\xgboost_winning_non_sensitive_validation.csv': 0},
 'checks': {'

{'stage': 'stage4j',
 'official_name': 'Stage 4J — XGBoost Feature Confirmation and Final Tuning',
 'created_at_utc': '2026-07-14T19:27:37.857116+00:00',
 'confirmation_fit_count': 3,
 'tuning_fit_count': 4,
 'selected_experiment_id': 'stage4j__tuning_a_baseline',
 'prediction_test_overlaps': {'artifacts\\predictions\\xgboost\\feature_confirmation\\confirmation_combined_v2.csv': 0,
  'artifacts\\predictions\\xgboost\\feature_confirmation\\confirmation_original.csv': 0,
  'artifacts\\predictions\\xgboost\\feature_confirmation\\confirmation_ratio_only_v2.csv': 0,
  'artifacts\\predictions\\xgboost\\final_selection\\tuning_a_baseline.csv': 0,
  'artifacts\\predictions\\xgboost\\final_selection\\tuning_b_safer.csv': 0,
  'artifacts\\predictions\\xgboost\\final_selection\\tuning_c_flexible.csv': 0,
  'artifacts\\predictions\\xgboost\\final_selection\\tuning_d_low_lr.csv': 0,
  'artifacts\\predictions\\xgboost\\final_selection\\xgboost_winning_non_sensitive_validation.csv': 0},
 'checks': {'

## Stage 4K — Final XGBoost Sensitive Comparison and Full-Train Models

### 32. Stage 4K Objective

Why: State the final bounded scope.

The code below reads compact saved evidence.

Interpretation: Stage 4K allows one sensitive Validation fit and two full-Train fits only.

Conclusion: No extra tuning or Test evaluation is part of this Gate.

In [33]:
print('Stage 4K reuses the non-sensitive winner, fits one sensitive Validation model, and saves two full-Train models.')

Stage 4K reuses the non-sensitive winner, fits one sensitive Validation model, and saves two full-Train models.


### 33. Stage 4J Gate Validation

Why: Recheck the second Gate dependency.

The code below reads compact saved evidence.

Interpretation: The Stage 4J Gate remains PASS.

Conclusion: Sensitive fitting starts only after configuration freeze.

In [34]:
show_json('artifacts/reports/stage4j_gate_verification.json')

{'stage': 'stage4j',
 'official_name': 'Stage 4J — XGBoost Feature Confirmation and Final Tuning',
 'created_at_utc': '2026-07-14T19:27:37.857116+00:00',
 'confirmation_fit_count': 3,
 'tuning_fit_count': 4,
 'selected_experiment_id': 'stage4j__tuning_a_baseline',
 'prediction_test_overlaps': {'artifacts\\predictions\\xgboost\\feature_confirmation\\confirmation_combined_v2.csv': 0,
  'artifacts\\predictions\\xgboost\\feature_confirmation\\confirmation_original.csv': 0,
  'artifacts\\predictions\\xgboost\\feature_confirmation\\confirmation_ratio_only_v2.csv': 0,
  'artifacts\\predictions\\xgboost\\final_selection\\tuning_a_baseline.csv': 0,
  'artifacts\\predictions\\xgboost\\final_selection\\tuning_b_safer.csv': 0,
  'artifacts\\predictions\\xgboost\\final_selection\\tuning_c_flexible.csv': 0,
  'artifacts\\predictions\\xgboost\\final_selection\\tuning_d_low_lr.csv': 0,
  'artifacts\\predictions\\xgboost\\final_selection\\xgboost_winning_non_sensitive_validation.csv': 0},
 'checks': {'

{'stage': 'stage4j',
 'official_name': 'Stage 4J — XGBoost Feature Confirmation and Final Tuning',
 'created_at_utc': '2026-07-14T19:27:37.857116+00:00',
 'confirmation_fit_count': 3,
 'tuning_fit_count': 4,
 'selected_experiment_id': 'stage4j__tuning_a_baseline',
 'prediction_test_overlaps': {'artifacts\\predictions\\xgboost\\feature_confirmation\\confirmation_combined_v2.csv': 0,
  'artifacts\\predictions\\xgboost\\feature_confirmation\\confirmation_original.csv': 0,
  'artifacts\\predictions\\xgboost\\feature_confirmation\\confirmation_ratio_only_v2.csv': 0,
  'artifacts\\predictions\\xgboost\\final_selection\\tuning_a_baseline.csv': 0,
  'artifacts\\predictions\\xgboost\\final_selection\\tuning_b_safer.csv': 0,
  'artifacts\\predictions\\xgboost\\final_selection\\tuning_c_flexible.csv': 0,
  'artifacts\\predictions\\xgboost\\final_selection\\tuning_d_low_lr.csv': 0,
  'artifacts\\predictions\\xgboost\\final_selection\\xgboost_winning_non_sensitive_validation.csv': 0},
 'checks': {'

### 34. Frozen Final Configuration

Why: Show the one configuration used in both modes.

The code below reads compact saved evidence.

Interpretation: Both modes use the ratio pack, log1p target, 982 trees, seed 42, CPU, and two threads.

Conclusion: The sensitive comparison changes only validated sensitive source columns.

In [35]:
final_config=show_json('artifacts/results/stage4/xgboost/final/stage4j_final_configuration.json')

{'stage': 'stage4j',
 'selected_experiment_id': 'stage4j__tuning_a_baseline',
 'selection_reason': 'Candidate D had 0.086 percent lower MAE, inside the 0.25 percent tie band; Candidate A uses about half as many trees and runtime.',
 'feature_pack': 'xgboost_sparse_v2_ratio',
 'feature_engineer_version': 'xgboost_feature_engineer_v2_20260714',
 'selected_proposals': ['applicant_to_estimated_tract_income_ratio'],
 'target_mode': 'log1p',
 'parameters': {'objective': 'reg:squarederror',
  'tree_method': 'hist',
  'device': 'cpu',
  'n_estimators': 982,
  'learning_rate': 0.05,
  'max_depth': 6,
  'min_child_weight': 10,
  'subsample': 0.8,
  'colsample_bytree': 0.8,
  'reg_alpha': 0.0,
  'reg_lambda': 5.0,
  'gamma': 0.0,
  'max_bin': 256,
  'random_state': 42,
  'n_jobs': 2,
  'eval_metric': 'mae'},
 'best_iteration': 982,
 'random_seed': 42,
 'execution_mode': 'cpu',
 'thread_count': 2,
 'encoding_policy': 'sparse one-hot plus training-fit frequency maps',
 'metrics': {'mae': 64.2219236

### 35. Reused Non-Sensitive Validation Result

Why: Prove that no Stage 4K non-sensitive refit occurred.

The code below reads compact saved evidence.

Interpretation: All Stage 4J hashes match and `refit_performed` is false.

Conclusion: The non-sensitive Validation result was reused exactly.

In [36]:
show_json('artifacts/reports/stage4k_non_sensitive_reuse_verification.json')

{'stage': 'stage4k',
 'source_stage': 'stage4j',
 'reused_experiment_id': 'stage4j__tuning_a_baseline',
 'checks': {'stage4j_gate_pass': True,
  'model_hash_matches': True,
  'native_hash_matches': True,
  'prediction_hash_matches': True,
  'manifest_matches_configuration': True,
  'prediction_rows_25000': True,
  'non_sensitive_mode': True,
  'fixed_iteration_982': True},
 'refit_performed': False,
 'status': 'PASS'}

{'stage': 'stage4k',
 'source_stage': 'stage4j',
 'reused_experiment_id': 'stage4j__tuning_a_baseline',
 'checks': {'stage4j_gate_pass': True,
  'model_hash_matches': True,
  'native_hash_matches': True,
  'prediction_hash_matches': True,
  'manifest_matches_configuration': True,
  'prediction_rows_25000': True,
  'non_sensitive_mode': True,
  'fixed_iteration_982': True},
 'refit_performed': False,
 'status': 'PASS'}

### 36. Single Sensitive Validation Fit

Why: Show the only sensitive Validation fit.

The code below reads compact saved evidence.

Interpretation: The single sensitive model has MAE 63.8138 on the same 25,000 Validation IDs.

Conclusion: The sensitive fit budget is exactly one.

In [37]:
sensitive=show_json('artifacts/checkpoints/stage4/xgboost/sensitive_validation.json'); display(sensitive['metrics'])

{'stage_id': 'stage4k',
 'stage_name': 'Stage 4K — Final XGBoost Sensitive Comparison and Full-Train Models',
 'experiment_id': 'stage4k__sensitive_validation',
 'job_type': 'validation',
 'feature_pack': 'xgboost_sparse_v2_ratio',
 'feature_engineer_version': 'xgboost_feature_engineer_v2_20260714',
 'selected_proposals': ['applicant_to_estimated_tract_income_ratio'],
 'target_mode': 'log1p',
 'sensitive_mode': 'with_sensitive',
 'parameters': {'objective': 'reg:squarederror',
  'tree_method': 'hist',
  'device': 'cpu',
  'n_estimators': 982,
  'learning_rate': 0.05,
  'max_depth': 6,
  'min_child_weight': 10,
  'subsample': 0.8,
  'colsample_bytree': 0.8,
  'reg_alpha': 0.0,
  'reg_lambda': 5.0,
  'gamma': 0.0,
  'max_bin': 256,
  'random_state': 42,
  'n_jobs': 2,
  'eval_metric': 'mae'},
 'fixed_or_best_iteration': 982,
 'random_seed': 42,
 'execution_mode': 'cpu',
 'thread_count': 2,
 'sample_name': 'final_selection',
 'sample_digest': '445a077b3fc1963bec38d8ba90384df2e1400e3897b25

{'mae': 63.8138342760758,
 'mse': 17041.189637274587,
 'rmse': 130.54190759014742,
 'mape_percent': 33.63717193258602,
 'r_squared': 0.6923616806904016,
 'rmsle': 0.38963898597834074,
 'rmsle_clipped_zero': 0.38963898597834074,
 'median_absolute_error': 37.40941443503141,
 'wape_percent': 25.849584096666128,
 'mean_signed_error': -15.668245545877996,
 'p90_absolute_error': 135.0046889060762,
 'negative_prediction_rate': 0.0,
 'mae_usd': 63813.8342760758,
 'rmse_usd': 130541.90759014741,
 'original_scale': True,
 'top_decile_mae': 207.15172914672627,
 'top_five_percent_mae': 299.82317317122454,
 'underestimation_rate': 0.53872,
 'overestimation_rate': 0.46128}

### 37. Controlled Sensitive Comparison

Why: Show with-minus-without differences as accuracy evidence only.

The code below reads compact saved evidence.

Interpretation: With-minus-without MAE is -0.4081; RMSE is -0.1021 and top-decile MAE is -2.2706.

Conclusion: This is controlled Validation accuracy evidence, not Test or fairness evidence.

In [38]:
show_csv('artifacts/results/stage4/xgboost/final/stage4k_sensitive_comparison.csv',20)

,metric,without_sensitive,with_sensitive,difference_with_minus_without
0,mae,64.221924,63.813834,-0.408089
1,mse,17067.861411,17041.189637,-26.671774
2,rmse,130.644026,130.541908,-0.102118
3,mape_percent,33.730802,33.637172,-0.093630
4,r_squared,0.691880,0.692362,0.000481
5,rmsle,0.391395,0.389639,-0.001756
6,rmsle_clipped_zero,0.391395,0.389639,-0.001756
7,median_absolute_error,37.598324,37.409414,-0.188910
8,wape_percent,26.014892,25.849584,-0.165308
9,mean_signed_error,-15.976944,-15.668246,0.308698


,metric,without_sensitive,with_sensitive,difference_with_minus_without
0,mae,64.221924,63.813834,-0.408089
1,mse,17067.861411,17041.189637,-26.671774
2,rmse,130.644026,130.541908,-0.102118
3,mape_percent,33.730802,33.637172,-0.093630
4,r_squared,0.691880,0.692362,0.000481
5,rmsle,0.391395,0.389639,-0.001756
6,rmsle_clipped_zero,0.391395,0.389639,-0.001756
7,median_absolute_error,37.598324,37.409414,-0.188910
8,wape_percent,26.014892,25.849584,-0.165308
9,mean_signed_error,-15.976944,-15.668246,0.308698


### 38. Full-Training Models

Why: Show both 399,788-row Train-only fits.

The code below reads compact saved evidence.

Interpretation: Both modes used all 399,788 saved Train rows and zero Test rows; fits took 31.43 and 34.37 seconds.

Conclusion: Exactly two sequential full-Train models were created.

In [39]:
display([show_json(f'artifacts/checkpoints/stage4/xgboost/full_train_{m}.json') for m in ['without_sensitive','with_sensitive']])

{'stage_id': 'stage4k',
 'stage_name': 'Stage 4K — Final XGBoost Sensitive Comparison and Full-Train Models',
 'experiment_id': 'stage4k__full_train_without_sensitive',
 'job_type': 'full_train',
 'feature_pack': 'xgboost_sparse_v2_ratio',
 'feature_engineer_version': 'xgboost_feature_engineer_v2_20260714',
 'selected_proposals': ['applicant_to_estimated_tract_income_ratio'],
 'target_mode': 'log1p',
 'sensitive_mode': 'without_sensitive',
 'parameters': {'objective': 'reg:squarederror',
  'tree_method': 'hist',
  'device': 'cpu',
  'n_estimators': 982,
  'learning_rate': 0.05,
  'max_depth': 6,
  'min_child_weight': 10,
  'subsample': 0.8,
  'colsample_bytree': 0.8,
  'reg_alpha': 0.0,
  'reg_lambda': 5.0,
  'gamma': 0.0,
  'max_bin': 256,
  'random_state': 42,
  'n_jobs': 2,
  'eval_metric': 'mae'},
 'fixed_or_best_iteration': 982,
 'random_seed': 42,
 'execution_mode': 'cpu',
 'thread_count': 2,
 'sample_name': 'full_train',
 'sample_digest': '5abb11436c4661f304c3b06ade1882cbfd7689c

{'stage_id': 'stage4k',
 'stage_name': 'Stage 4K — Final XGBoost Sensitive Comparison and Full-Train Models',
 'experiment_id': 'stage4k__full_train_with_sensitive',
 'job_type': 'full_train',
 'feature_pack': 'xgboost_sparse_v2_ratio',
 'feature_engineer_version': 'xgboost_feature_engineer_v2_20260714',
 'selected_proposals': ['applicant_to_estimated_tract_income_ratio'],
 'target_mode': 'log1p',
 'sensitive_mode': 'with_sensitive',
 'parameters': {'objective': 'reg:squarederror',
  'tree_method': 'hist',
  'device': 'cpu',
  'n_estimators': 982,
  'learning_rate': 0.05,
  'max_depth': 6,
  'min_child_weight': 10,
  'subsample': 0.8,
  'colsample_bytree': 0.8,
  'reg_alpha': 0.0,
  'reg_lambda': 5.0,
  'gamma': 0.0,
  'max_bin': 256,
  'random_state': 42,
  'n_jobs': 2,
  'eval_metric': 'mae'},
 'fixed_or_best_iteration': 982,
 'random_seed': 42,
 'execution_mode': 'cpu',
 'thread_count': 2,
 'sample_name': 'full_train',
 'sample_digest': '5abb11436c4661f304c3b06ade1882cbfd7689c8befcd

[{'stage_id': 'stage4k',
  'stage_name': 'Stage 4K — Final XGBoost Sensitive Comparison and Full-Train Models',
  'experiment_id': 'stage4k__full_train_without_sensitive',
  'job_type': 'full_train',
  'feature_pack': 'xgboost_sparse_v2_ratio',
  'feature_engineer_version': 'xgboost_feature_engineer_v2_20260714',
  'selected_proposals': ['applicant_to_estimated_tract_income_ratio'],
  'target_mode': 'log1p',
  'sensitive_mode': 'without_sensitive',
  'parameters': {'objective': 'reg:squarederror',
   'tree_method': 'hist',
   'device': 'cpu',
   'n_estimators': 982,
   'learning_rate': 0.05,
   'max_depth': 6,
   'min_child_weight': 10,
   'subsample': 0.8,
   'colsample_bytree': 0.8,
   'reg_alpha': 0.0,
   'reg_lambda': 5.0,
   'gamma': 0.0,
   'max_bin': 256,
   'random_state': 42,
   'n_jobs': 2,
   'eval_metric': 'mae'},
  'fixed_or_best_iteration': 982,
  'random_seed': 42,
  'execution_mode': 'cpu',
  'thread_count': 2,
  'sample_name': 'full_train',
  'sample_digest': '5abb1143

### 39. Final Model Manifests

Why: Show complete bundles, native models, and provenance.

The code below reads compact saved evidence.

Interpretation: Each manifest records the full bundle, native JSON, split/source hashes, versions, and reference predictions.

Conclusion: Bare estimators were not used as final artifacts.

In [40]:
display([show_json(f'artifacts/manifests/stage4/xgboost/stage4k_final_model_manifest_{m}.json') for m in ['without_sensitive','with_sensitive']])

### 40. Clean-Process Reload

Why: Confirm both final reload and sparse checks.

The code below reads compact saved evidence.

Interpretation: Both modes match bundle, native, and reference predictions; sparse shapes are 161 and 197.

Conclusion: Both final models reload safely in clean processes.

In [41]:
display([show_json(f'artifacts/reports/stage4k_final_reload_{m}.json') for m in ['without_sensitive','with_sensitive']])

{'model_path': 'artifacts\\models\\xgboost\\final\\xgboost_final_without_sensitive.joblib',
 'native_model_path': 'artifacts\\models\\xgboost\\final\\xgboost_final_without_sensitive.json',
 'prediction_path': 'artifacts\\predictions\\xgboost\\final\\xgboost_full_train_reference_without_sensitive.csv',
 'rows': 500,
 'finite_predictions': True,
 'output_length_matches': True,
 'source_frame_unchanged': True,
 'reference_prediction_match': True,
 'native_prediction_match': True,
 'inverse_target_verified': True,
 'native_model_loaded': True,
 'sparse_transformation': True,
 'sparse_shape': [500, 161],
 'status': 'PASS'}

{'model_path': 'artifacts\\models\\xgboost\\final\\xgboost_final_with_sensitive.joblib',
 'native_model_path': 'artifacts\\models\\xgboost\\final\\xgboost_final_with_sensitive.json',
 'prediction_path': 'artifacts\\predictions\\xgboost\\final\\xgboost_full_train_reference_with_sensitive.csv',
 'rows': 500,
 'finite_predictions': True,
 'output_length_matches': True,
 'source_frame_unchanged': True,
 'reference_prediction_match': True,
 'native_prediction_match': True,
 'inverse_target_verified': True,
 'native_model_loaded': True,
 'sparse_transformation': True,
 'sparse_shape': [500, 197],
 'status': 'PASS'}

[{'model_path': 'artifacts\\models\\xgboost\\final\\xgboost_final_without_sensitive.joblib',
  'native_model_path': 'artifacts\\models\\xgboost\\final\\xgboost_final_without_sensitive.json',
  'prediction_path': 'artifacts\\predictions\\xgboost\\final\\xgboost_full_train_reference_without_sensitive.csv',
  'rows': 500,
  'finite_predictions': True,
  'output_length_matches': True,
  'source_frame_unchanged': True,
  'reference_prediction_match': True,
  'native_prediction_match': True,
  'inverse_target_verified': True,
  'native_model_loaded': True,
  'sparse_transformation': True,
  'sparse_shape': [500, 161],
  'status': 'PASS'},
 {'model_path': 'artifacts\\models\\xgboost\\final\\xgboost_final_with_sensitive.joblib',
  'native_model_path': 'artifacts\\models\\xgboost\\final\\xgboost_final_with_sensitive.json',
  'prediction_path': 'artifacts\\predictions\\xgboost\\final\\xgboost_full_train_reference_with_sensitive.csv',
  'rows': 500,
  'finite_predictions': True,
  'output_length_

### 41. Final Gain and Weight Importance

Why: Show final importance for both modes.

The code below reads compact saved evidence.

Interpretation: Final gain/weight tables exist for both modes; lien and income Features remain leading signals.

Conclusion: Importance is available for comparison without a causal claim.

In [42]:
display({m:pd.read_csv(ROOT/f'artifacts/results/stage4/xgboost/final/stage4k_importance_top20_{m}.csv').head(20) for m in ['without_sensitive','with_sensitive']})

{'without_sensitive':                         source_feature          gain  weight    total_gain
 0                     lien_status_name  10724.256592   277.0  1.043573e+06
 1                applicant_income_000s    189.404236  2463.0  4.665026e+05
 2   estimated_tract_family_income_000s     61.006042  2240.0  1.366535e+05
 3    applicant_vs_area_income_gap_000s     42.157837  1481.0  6.243576e+04
 4            purpose_lien_status_group   3725.659066   858.0  5.829849e+04
 5                            us_region    282.435422   719.0  5.760265e+04
 6      applicant_income_to_area_income     31.480631  1760.0  5.540591e+04
 7              state_lien_status_group     31.060041  1717.0  5.333009e+04
 8             hud_median_family_income     21.586918  2300.0  4.964991e+04
 9               property_purpose_group    169.730597   854.0  4.161191e+04
 10         occupancy_lien_status_group    264.911000   730.0  3.417184e+04
 11                owner_occupancy_name    152.439615   426.0  2.78

### 42. Final Native SHAP

Why: Show same-ID bounded native SHAP for both modes.

The code below reads compact saved evidence.

Interpretation: Both modes use the same 300 Final Selection IDs for native contributions.

Conclusion: SHAP is bounded, associative, and never used as a model Feature.

In [43]:
display({m:pd.read_csv(ROOT/f'artifacts/results/stage4/xgboost/final/stage4k_shap_top20_{m}.csv').head(20) for m in ['without_sensitive','with_sensitive']}); show_csv('artifacts/manifests/stage4/xgboost/stage4k_shap_row_ids.csv',5)

{'without_sensitive':                         source_feature  mean_absolute_shap
 0                     lien_status_name            0.237359
 1                applicant_income_000s            0.222693
 2   estimated_tract_family_income_000s            0.119719
 3                            us_region            0.096444
 4               property_purpose_group            0.086179
 5             hud_median_family_income            0.053734
 6                           state_name            0.052481
 7          occupancy_lien_status_group            0.051531
 8              state_lien_status_group            0.051516
 9                       loan_type_name            0.040596
 10   applicant_vs_area_income_gap_000s            0.034518
 11                          msamd_name            0.029822
 12            agency_lien_status_group            0.028729
 13                         agency_name            0.022855
 14     applicant_income_to_area_income            0.022639
 15                

### 43. Previous-Stage Reference

Why: Keep unlike evaluation designs clearly labelled.

The code below reads compact saved evidence.

Interpretation: Stage 2/3 rows are OOF, while CatBoost, LightGBM, and XGBoost rows are Final Selection Validation.

Conclusion: Only the future locked Test evaluation can make a common final comparison.

In [44]:
show_csv('artifacts/results/stage4/xgboost/final/stage4k_previous_stage_reference.csv',10); print('Stage 2 and Stage 3 are OOF. Boosting rows use Final Selection Validation. The locked Test Set will provide the common final comparison.')

,stage,model,evaluation_design,evaluation_rows,sensitive_mode,mae,rmse,rmsle,r_squared,runtime_seconds
0,Stage 2,lasso,OOF,399788,with_sensitive,72.126971,248.003710,0.440883,0.390727,84.581682
1,Stage 3,hist_gradient_boosting,OOF,399788,with_sensitive,64.054533,249.196569,0.394470,0.384852,48.689495
2,Stage 4D-E,CatBoost,Final Selection Validation,25000,without_sensitive,63.376185,130.272187,0.390619,0.693632,320.629190
3,Stage 4D-E,CatBoost,Final Selection Validation,25000,with_sensitive,63.273525,131.685736,0.390349,0.686947,404.242927
4,Stage 4G-H,LightGBM,Final Selection Validation,25000,without_sensitive,64.456751,131.956170,0.445551,0.685660,14.205618
5,Stage 4G-H,LightGBM,Final Selection Validation,25000,with_sensitive,64.332196,132.464892,0.448031,0.683231,15.309392
6,Stage 4J-K,XGBoost,Final Selection Validation,25000,without_sensitive,64.221924,130.644026,0.391395,0.691880,42.563350
7,Stage 4J-K,XGBoost,Final Selection Validation,25000,with_sensitive,63.813834,130.541908,0.389639,0.692362,47.259242


Stage 2 and Stage 3 are OOF. Boosting rows use Final Selection Validation. The locked Test Set will provide the common final comparison.


### 44. XGBoost Artifact Summary

Why: Show artifact counts, Test exclusion, and runtime.

The code below reads compact saved evidence.

Interpretation: There are 16 XGBoost prediction files with zero Test overlap; accounted runtime is 530.88 seconds.

Conclusion: The artifact set is complete and inside the 210-minute limit.

In [45]:
show_json('artifacts/reports/stage4k_artifact_summary.json')

{'stage': 'stage4k',
 'created_at_utc': '2026-07-14T19:35:22.985573+00:00',
 'xgboost_prediction_files': 16,
 'prediction_test_overlaps': {'artifacts\\predictions\\xgboost\\feature_confirmation\\confirmation_combined_v2.csv': 0,
  'artifacts\\predictions\\xgboost\\feature_confirmation\\confirmation_original.csv': 0,
  'artifacts\\predictions\\xgboost\\feature_confirmation\\confirmation_ratio_only_v2.csv': 0,
  'artifacts\\predictions\\xgboost\\final\\xgboost_full_train_reference_without_sensitive.csv': 0,
  'artifacts\\predictions\\xgboost\\final\\xgboost_full_train_reference_with_sensitive.csv': 0,
  'artifacts\\predictions\\xgboost\\final_selection\\sensitive_validation.csv': 0,
  'artifacts\\predictions\\xgboost\\final_selection\\tuning_a_baseline.csv': 0,
  'artifacts\\predictions\\xgboost\\final_selection\\tuning_b_safer.csv': 0,
  'artifacts\\predictions\\xgboost\\final_selection\\tuning_c_flexible.csv': 0,
  'artifacts\\predictions\\xgboost\\final_selection\\tuning_d_low_lr.csv'

{'stage': 'stage4k',
 'created_at_utc': '2026-07-14T19:35:22.985573+00:00',
 'xgboost_prediction_files': 16,
 'prediction_test_overlaps': {'artifacts\\predictions\\xgboost\\feature_confirmation\\confirmation_combined_v2.csv': 0,
  'artifacts\\predictions\\xgboost\\feature_confirmation\\confirmation_original.csv': 0,
  'artifacts\\predictions\\xgboost\\feature_confirmation\\confirmation_ratio_only_v2.csv': 0,
  'artifacts\\predictions\\xgboost\\final\\xgboost_full_train_reference_without_sensitive.csv': 0,
  'artifacts\\predictions\\xgboost\\final\\xgboost_full_train_reference_with_sensitive.csv': 0,
  'artifacts\\predictions\\xgboost\\final_selection\\sensitive_validation.csv': 0,
  'artifacts\\predictions\\xgboost\\final_selection\\tuning_a_baseline.csv': 0,
  'artifacts\\predictions\\xgboost\\final_selection\\tuning_b_safer.csv': 0,
  'artifacts\\predictions\\xgboost\\final_selection\\tuning_c_flexible.csv': 0,
  'artifacts\\predictions\\xgboost\\final_selection\\tuning_d_low_lr.csv'

### 45. Final Verification

Why: Show the strongest verification available for this run.

The code below reads compact saved evidence.

Interpretation: The cell shows final verification when available; before review it shows the clearly named pre-review report.

Conclusion: A pre-review PASS is not treated as final completion.

In [46]:
verify_path='artifacts/reports/stage4k_verification.json' if (ROOT/'artifacts/reports/stage4k_verification.json').is_file() else 'artifacts/reports/stage4k_pre_review_verification.json'; show_json(verify_path)

{'stage': 'stage4k',
 'official_name': 'Stage 4K — Final XGBoost Sensitive Comparison and Full-Train Models',
 'created_at_utc': '2026-07-14T19:45:15.970310+00:00',
 'checks': {'all_prior_stages_pass': True,
  'stage4i_gate_pass': True,
  'stage4j_gate_pass': True,
  'test_overlap_zero_all_xgboost_predictions': True,
  'non_sensitive_winner_reused': True,
  'exactly_one_sensitive_validation_fit': True,
  'same_frozen_configuration_both_modes': True,
  'exactly_two_full_train_fits': True,
  'full_train_rows_399788_each': True,
  'two_complete_bundles_and_native_models': True,
  'two_clean_reload_tests_pass': True,
  'sparse_transformation_valid': True,
  'final_importance_both_modes': True,
  'final_shap_both_modes_same_300_ids': True,
  'registry_ids_unique': True,
  'prior_registry_bytes_preserved': True,
  'protected_hashes_unchanged': True,
  'runtime_within_limit': True,
  'required_heavy_artifacts_complete': True,
  'complete_notebook_run_pass': True,
  'cache_only_notebook_run_pa

{'stage': 'stage4k',
 'official_name': 'Stage 4K — Final XGBoost Sensitive Comparison and Full-Train Models',
 'created_at_utc': '2026-07-14T19:45:15.970310+00:00',
 'checks': {'all_prior_stages_pass': True,
  'stage4i_gate_pass': True,
  'stage4j_gate_pass': True,
  'test_overlap_zero_all_xgboost_predictions': True,
  'non_sensitive_winner_reused': True,
  'exactly_one_sensitive_validation_fit': True,
  'same_frozen_configuration_both_modes': True,
  'exactly_two_full_train_fits': True,
  'full_train_rows_399788_each': True,
  'two_complete_bundles_and_native_models': True,
  'two_clean_reload_tests_pass': True,
  'sparse_transformation_valid': True,
  'final_importance_both_modes': True,
  'final_shap_both_modes_same_300_ids': True,
  'registry_ids_unique': True,
  'prior_registry_bytes_preserved': True,
  'protected_hashes_unchanged': True,
  'runtime_within_limit': True,
  'required_heavy_artifacts_complete': True,
  'complete_notebook_run_pass': True,
  'cache_only_notebook_run_pa

### 46. XGBoost Track Completion Note

Why: State completion honestly and keep Test evaluation out of scope.

The code below reads compact saved evidence.

Interpretation: Completion becomes true only when `stage4k_verification.json` exists and reports PASS.

Conclusion: The next step is exposed only after final verification; Test evaluation is not started here.

In [47]:
final_path=ROOT/'artifacts/reports/stage4k_verification.json'
final_status=json.loads(final_path.read_text(encoding='utf-8'))['status'] if final_path.is_file() else None
final_complete=final_status=='PASS'
print({'cache_only':CACHE_ONLY,'final_verification_status':final_status,'final_complete':final_complete,'test_evaluation_started':False,'next_step':'Begin Final Model Integration and Locked Test Evaluation.' if final_complete else None})

{'cache_only': True, 'final_verification_status': 'PASS', 'final_complete': True, 'test_evaluation_started': False, 'next_step': 'Begin Final Model Integration and Locked Test Evaluation.'}
